src/main/scala/org/broadinstitute/hail/expr/AST.scala

```diff
@@ -227,8 +361,502 @@ sealed abstract class AST(pos: Position, subexprs: Array[AST] = Array.empty) {
       parseError(msg)
     subexprs.foreach(_.errorIf[T](msg))
   }
+
+  def compile(): CM[Code[AnyRef]]
+
+  def compileAggregator(): CMCodeCPS[AnyRef]
+
+  def run(ec: EvalContext): () => AnyRef = this.compile().run(ec)
+
+  def runAggregator(ec: EvalContext): CPS[Any] = {
+    val names = ec.st.toSeq
+      .sortBy { case (nanme, (i, _)) => i }
+      .map { case (name, _) => name }
+    val values = ec.a.asInstanceOf[mutable.ArrayBuffer[AnyRef]]
+
+    val idx = ec.a.length
+    val localA = ec.a
+    localA += null
+
+    val f = (this.compileAggregator() { (me: Code[AnyRef]) => for (
+      vs <- CM.initialValueArray();
+      k = Code.checkcast[AnyRef => Unit](vs.invoke[Int, AnyRef]("apply", idx))
+      // actually returns Unit, but invoke will not find the apply method if we enter Unit
+    ) yield k.invoke[AnyRef, AnyRef]("apply", me).asInstanceOf[Code[Unit]]
+      // FIXME: how do I avoid this cast ??
+    }).asInstanceOf[CM[Code[AnyRef]]].runWithDelayedValues(names, ec)
+
+    { (k: (Any => Unit)) =>
+      localA(idx) = k
+      f(values)
+    }
+  }
 }
 
+object preCM {
+  case class E(m: Map[String, Code[AnyRef]], fb: Function2Builder[Array[AnyRef],mutable.ArrayBuffer[AnyRef],AnyRef])
+  case class S(fa: Array[AnyRef], ec: EvalContext)
+
+  def emptyE = E(Map[String, Code[AnyRef]](), new Function2Builder[Array[AnyRef],mutable.ArrayBuffer[AnyRef],AnyRef]())
+  def emptyS(ec: EvalContext) = S(Array[AnyRef](), ec)
+}
+
+import preCM._
+
+case class CM[+T](mt: (E, S) => (T, S)) {
```

Comment:
This is the "Code Monad", but it doesn't build `Code[T]` it just helps pipe around things like environments and foreign functions. Ideally it would *actually* build a `Code[T]`, but I haven't taken the time to figure that out. It'd be nice to do things like:

```
for (
  x <- newLocal(foo);
  y <- f.invoke[AnyRef, AnyRef](x)
  res <- y.ifNull(..., ...)
) yield res
```